File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    // Init res
    // Init left and right pointer
    let res = nums[0];
    let left = 0;
    let right = nums.length - 1;
    
    // while left <= right:
    //  EXP: If the subarray is already sorted, update the result and break
    //  if nums[left] < nums[right] :
    //      res is min between res and nums[left]
    //      break out of the loop
    //  calculate mid as Math.floor of left + right / 2
    //  EXP: Determine which side to search on because of the rotation
    //       If nums[mid] >= nums[left] means the smaller half is on the right side
    //       otherwise the smaller half is on the left side
    //  if nums[mid] >= nums[left]:
    //      left = mid + 1;
    //  else :
    //      right = mid - 1;
    
    while (left <= right) {
        if (nums[left] < nums[right]) {
            res = Math.min(res,nums[left]);
            break;
        }
        
        let mid = Math.floor((left + right)/2);
        res = Math.min(res,nums[mid]);
        if (nums[mid] >= nums[left]) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
        
    }
    
    // return res
    return res;
    // Time complexity : O(n) because each iteration of binary search reduces search by half
    // Space complexity : O(1) because we use constant amount of space 
};

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
class Solution:
    # Time complexity: O(logn)
    # Space complexity: O(1)
    def findMin(self, nums: List[int]) -> int:
        # init left pointer first index of num
        left = 0
        # init right pointer last index of num
        right = len(nums) -1
        # init res = first element of nums
        res = nums[0]
        # while left less than or equal to right pointer:
        #   if nums at left < nums at right (in sort section):
        #       update the res with nums at left
        #       break for loop
        #   calculate mid posititon
        #   update the res with nums at mid
        #   if nums at mid >= nums at left:
        #       move the left pointer to mid + 1
        #   else:
        #       move the right pointer to mid - 1
        
        # return res
        
        while left <= right:
            if nums[left] < nums[right]:
                res = min(res,nums[left])
                break
            mid = (left + right) // 2
            print(mid)
            res = min(res,nums[mid])
            
            if nums[mid] >= nums[left]:
                left = mid + 1
            else:
                right = mid - 1
        return res